nemo/collections/asr/models/label_models.py

```diff
@@ -319,97 +318,58 @@ def training_step(self, batch, batch_idx):
-            self.log('val_epoch_accuracy_top@{}'.format(top_k), score)
-
-        self.log('val_acc_macro', macro_accuracy_score, sync_dist=True)
-        self.log('val_auroc', auroc_score, sync_dist=True)
+            self.log(f'{tag}_acc_micro_top@{top_k}', score, sync_dist=True)
+        self.log(f'{tag}_acc_macro', macro_accuracy_score, sync_dist=True)
 
         return {
-            'val_loss': val_loss_mean,
-            'val_acc_top_k': topk_scores,
-            'val_acc_macro': macro_accuracy_score,
-            'val_auroc': auroc_score,
+            f'{tag}_loss': loss_mean,
+            f'{tag}_acc_top_k': topk_scores,
+            f'{tag}_acc_macro': macro_accuracy_score,
         }
 
     def test_step(self, batch, batch_idx, dataloader_idx: int = 0):
-        audio_signal, audio_signal_len, labels, _ = batch
-        logits, _ = self.forward(input_signal=audio_signal, input_signal_length=audio_signal_len)
-        loss_value = self.loss(logits=logits, labels=labels)
-        acc_top_k = self._accuracy(logits=logits, labels=labels)
-        correct_counts, total_counts = self._accuracy.correct_counts_k, self._accuracy.total_counts_k
-        self._macro_accuracy.update(preds=logits, target=labels)
-        self._auroc.update(preds=logits, target=labels)
-
-        return {
-            'test_loss': loss_value,
-            'test_correct_counts': correct_counts,
-            'test_total_counts': total_counts,
-            'test_acc_top_k': acc_top_k,
-        }
+        return self.validation_step(batch, batch_idx, dataloader_idx, 'test')
 
     def multi_test_epoch_end(self, outputs, dataloader_idx: int = 0):
-        test_loss_mean = torch.stack([x['test_loss'] for x in outputs]).mean()
-        correct_counts = torch.stack([x['test_correct_counts'] for x in outputs]).sum(axis=0)
-        total_counts = torch.stack([x['test_total_counts'] for x in outputs]).sum(axis=0)
-
-        self._accuracy.correct_counts_k = correct_counts
-        self._accuracy.total_counts_k = total_counts
-        topk_scores = self._accuracy.compute()
-        macro_accuracy_score = self._macro_accuracy.compute()
-        auroc_score = self._auroc.compute()
-
-        self._accuracy.reset()
-        self._macro_accuracy.reset()
-        self._auroc.reset()
-
-        logging.info("test_loss: {:.3f}".format(test_loss_mean))
-        self.log('test_loss', test_loss_mean)
-        for top_k, score in zip(self._accuracy.top_k, topk_scores):
-            self.log('test_epoch_accuracy_top@{}'.format(top_k), score)
-
-        self.log('test_acc_macro', macro_accuracy_score, sync_dist=True)
-        self.log('test_auroc', auroc_score, sync_dist=True)
-
-        return {
-            'test_loss': test_loss_mean,
-            'test_acc_top_k': topk_scores,
-            'test_acc_macro': macro_accuracy_score,
-            'test_auroc': auroc_score,
-        }
+        return self.multi_validation_epoch_end(outputs, dataloader_idx, 'test')
```

Comment:
Not sure if it's a good idea to merge val and test here

Reply:
what do you think? @nithinraok

Reply:
we can merge, why do you think its not a good idea?

Reply:
@nithinraok  just worried it would break things such eval on multi gpu and reduce flexibility, but I guess I think too much.  Then let's merge.

Reply:
I dont think since its re-use of function, but @fayejf or @stevehuang52 , can you run an experiment and see this working as expected before merging? We have a unit test for training but do not check validation and test time.

Reply:
I can do it, but do you have any examples of how to do it quickly? I think it will work correctly, and ASR models are also doing it the same way: https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/models/ctc_models.py#L650

Reply:
@fayejf can you provide steve with some lang id examples. Notice the validation time before and after the change.

Reply:
let me run it really quick.

Reply:
What might be cleaner is to separate out into a function if we're worried about using the validation hook within the test hook:

```python
def calculate_metrics(self, outputs, dataloader_idx, tag):
    ...

def multi_test_epoch_end(self, outputs, dataloader_idx):
   return self.calculate_metrics(outputs, dataloader_idx, 'test')

def multi_validation_epoch_end(self, outputs, dataloader_idx):
   return self.calculate_metrics(outputs, dataloader_idx, 'val')

```

Reply:
yes, this would be neat.